File: cogs/extralife.py
"""Extra-Life charity integration cog."""

import discord
from discord.ext import commands, tasks
from discord import app_commands
import aiohttp
import asyncio
from typing import Optional, Dict, Any
from datetime import datetime, timedelta
from config import Config
from utils.database import DataManager

class ExtraLife(commands.Cog):
    """Extra-Life charity event integration."""
    
    def __init__(self, bot):
        self.bot = bot
        self.data_manager = DataManager()
        self.session = None
        self.announcement_channel = None
        self.pinned_message = None
        
        # Start background tasks
        if Config.EXTRALIFE_TEAM_ID or Config.EXTRALIFE_PARTICIPANT_ID:
            self.update_stats.start()
    
    async def cog_load(self):
        """Initialize HTTP session when cog loads."""
        self.session = aiohttp.ClientSession()
    
    async def cog_unload(self):
        """Clean up HTTP session when cog unloads."""
        if self.session:
            await self.session.close()
        self.update_stats.cancel()
    
    @app_commands.command(name="extralife", description="Extra-Life charity integration commands")
    @app_commands.describe(
        action="Action to perform",
        channel="Channel for announcements (admin only)"
    )
    async def extralife_command(
        self,
        interaction: discord.Interaction,
        action: str = "stats",
        channel: Optional[discord.TextChannel] = None
    ):
        """Main Extra-Life command."""
        if action == "stats":
            await self.show_stats(interaction)
        elif action == "setup":
            await self.setup_announcements(interaction, channel)
        elif action == "announce":
            await self.post_announcement(interaction)
        elif action == "refresh":
            await self.refresh_stats(interaction)
        else:
            await interaction.response.send_message(
                "❌ Invalid action. Use: `stats`, `setup`, `announce`, or `refresh`",
                ephemeral=True
            )
    
    async def show_stats(self, interaction: discord.Interaction):
        """Show current Extra-Life statistics."""
        await interaction.response.defer()
        
        try:
            # Try to get cached data first
            cached_data = self.data_manager.load_extralife_cache()
            if cached_data:
                data = cached_data
            else:
                data = await self.fetch_extralife_data()
                if data:
                    self.data_manager.save_extralife_cache(data)
            
            if not data:
                await interaction.followup.send("❌ Unable to fetch Extra-Life data. Check configuration.", ephemeral=True)
                return
            
            embed = self.create_stats_embed(data)
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            await interaction.followup.send(f"❌ Error fetching stats: {str(e)}", ephemeral=True)
    
    async def setup_announcements(self, interaction: discord.Interaction, channel: Optional[discord.TextChannel]):
        """Setup announcement channel (admin only)."""
        if not interaction.user.guild_permissions.administrator:
            await interaction.response.send_message("❌ Administrator permissions required.", ephemeral=True)
            return
        
        if not channel:
            channel = interaction.channel
        
        self.announcement_channel = channel
        
        embed = discord.Embed(
            title="🎮 Extra-Life Setup Complete",
            description=f"Announcements will be posted in {channel.mention}",
            color=discord.Color.green()
        )
        
        await interaction.response.send_message(embed=embed)
    
    async def post_announcement(self, interaction: discord.Interaction):
        """Post Extra-Life event announcement."""
        if not self.announcement_channel:
            await interaction.response.send_message(
                "❌ No announcement channel set. Use `/extralife setup` first.",
                ephemeral=True
            )
            return
        
        await interaction.response.defer()
        
        try:
            data = await self.fetch_extralife_data()
            if not data:
                await interaction.followup.send("❌ Unable to fetch Extra-Life data.", ephemeral=True)
                return
            
            embed = self.create_announcement_embed(data)
            
            # Post to announcement channel
            message = await self.announcement_channel.send(embed=embed)
            
            # Try to pin the message
            try:
                await message.pin()
                self.pinned_message = message
            except discord.Forbidden:
                pass  # No permission to pin
            
            await interaction.followup.send(f"✅ Announcement posted in {self.announcement_channel.mention}")
            
        except Exception as e:
            await interaction.followup.send(f"❌ Error posting announcement: {str(e)}", ephemeral=True)
    
    async def refresh_stats(self, interaction: discord.Interaction):
        """Manually refresh Extra-Life statistics."""
        await interaction.response.defer()
        
        try:
            data = await self.fetch_extralife_data()
            if data:
                self.data_manager.save_extralife_cache(data)
                embed = self.create_stats_embed(data)
                await interaction.followup.send("✅ Stats refreshed!", embed=embed)
            else:
                await interaction.followup.send("❌ Unable to fetch fresh data.", ephemeral=True)
                
        except Exception as e:
            await interaction.followup.send(f"❌ Error refreshing: {str(e)}", ephemeral=True)
    
    async def fetch_extralife_data(self) -> Optional[Dict[str, Any]]:
        """Fetch data from Extra-Life API."""
        if not self.session:
            self.session = aiohttp.ClientSession()
        
        data = {}
        
        try:
            # Fetch team data
            if Config.EXTRALIFE_TEAM_ID:
                team_url = Config.get_extralife_team_url()
                async with self.session.get(team_url) as response:
                    if response.status == 200:
                        data['team'] = await response.json()
            
            # Fetch participant data
            if Config.EXTRALIFE_PARTICIPANT_ID:
                participant_url = Config.get_extralife_participant_url()
                async with self.session.get(participant_url) as response:
                    if response.status == 200:
                        data['participant'] = await response.json()
            
            return data if data else None
            
        except Exception as e:
            print(f"Error fetching Extra-Life data: {e}")
            return None
    
    def create_stats_embed(self, data: Dict[str, Any]) -> discord.Embed:
        """Create embed showing current statistics."""
        embed = discord.Embed(
            title="🎮 Extra-Life Statistics",
            color=discord.Color.blue(),
            timestamp=datetime.utcnow()
        )
        
        # Team stats
        if 'team' in data:
            team = data['team']
            embed.add_field(
                name="👥 Team Stats",
                value=(
                    f"**Name:** {team.get('name', 'Unknown')}\n"
                    f"**Goal:** ${team.get('fundraisingGoal', 0):,.2f}\n"
                    f"**Raised:** ${team.get('sumDonations', 0):,.2f}\n"
                    f"**Members:** {team.get('numMembers', 0)}"
                ),
                inline=True
            )
        
        # Participant stats
        if 'participant' in data:
            participant = data['participant']
            embed.add_field(
                name="🎯 Participant Stats",
                value=(
                    f"**Name:** {participant.get('displayName', 'Unknown')}\n"
                    f"**Goal:** ${participant.get('fundraisingGoal', 0):,.2f}\n"
                    f"**Raised:** ${participant.get('sumDonations', 0):,.2f}\n"
                    f"**Donors:** {participant.get('numDonations', 0)}"
                ),
                inline=True
            )
        
        # Progress calculation
        if 'team' in data:
            team = data['team']
            goal = team.get('fundraisingGoal', 1)
            raised = team.get('sumDonations', 0)
            progress = min(100, (raised / goal) * 100) if goal > 0 else 0
            
            embed.add_field(
                name="📊 Progress",
                value=f"**{progress:.1f}%** of goal reached",
                inline=False
            )
        
        embed.set_footer(text="Data from Extra-Life API • Updates every 5 minutes")
        
        return embed
    
    def create_announcement_embed(self, data: Dict[str, Any]) -> discord.Embed:
        """Create announcement embed for events."""
        embed = discord.Embed(
            title="🎮 Extra-Life Charity Event!",
            description=(
                "Join us in supporting Children's Miracle Network Hospitals! "
                "Every donation helps kids in our local hospitals."
            ),
            color=discord.Color.gold()
        )
        
        if 'team' in data:
            team = data['team']
            embed.add_field(
                name="🎯 Our Goal",
                value=f"${team.get('fundraisingGoal', 0):,.2f}",
                inline=True
            )
            embed.add_field(
                name="💰 Raised So Far",
                value=f"${team.get('sumDonations', 0):,.2f}",
                inline=True
            )
            embed.add_field(
                name="👥 Team Members",
                value=f"{team.get('numMembers', 0)} heroes",
                inline=True
            )
        
        embed.add_field(
            name="🎮 How to Help",
            value=(
                "• Donate directly to our team\n"
                "• Share our fundraising page\n"
                "• Join our gaming events\n"
                "• Spread the word!"
            ),
            inline=False
        )
        
        if Config.EXTRALIFE_TEAM_ID:
            team_page = f"https://www.extra-life.org/team/{Config.EXTRALIFE_TEAM_ID}"
            embed.add_field(
                name="🔗 Donation Link",
                value=f"[Visit Our Team Page]({team_page})",
                inline=False
            )
        
        embed.set_footer(text="Thank you for supporting children's hospitals! 💙")
        
        return embed
    
    @tasks.loop(minutes=5)
    async def update_stats(self):
        """Background task to update statistics."""
        try:
            data = await self.fetch_extralife_data()
            if data:
                self.data_manager.save_extralife_cache(data)
                
                # Update pinned message if it exists
                if self.pinned_message and self.announcement_channel:
                    try:
                        embed = self.create_announcement_embed(data)
                        await self.pinned_message.edit(embed=embed)
                    except (discord.NotFound, discord.Forbidden):
                        self.pinned_message = None
                        
        except Exception as e:
            print(f"Error in background stats update: {e}")
    
    @update_stats.before_loop
    async def before_update_stats(self):
        """Wait for bot to be ready before starting background task."""
        await self.bot.wait_until_ready()
    
    @app_commands.command(name="extralife-help", description="Show Extra-Life integration help")
    async def extralife_help(self, interaction: discord.Interaction):
        """Show help for Extra-Life commands."""
        embed = discord.Embed(
            title="🎮 Extra-Life Integration Help",
            description="Commands for charity event support",
            color=discord.Color.blue()
        )
        
        embed.add_field(
            name="📊 `/extralife stats`",
            value="Show current fundraising statistics",
            inline=False
        )
        
        embed.add_field(
            name="⚙️ `/extralife setup`",
            value="Setup announcement channel (admin only)",
            inline=False
        )
        
        embed.add_field(
            name="📢 `/extralife announce`",
            value="Post event announcement with current stats",
            inline=False
        )
        
        embed.add_field(
            name="🔄 `/extralife refresh`",
            value="Manually refresh statistics from API",
            inline=False
        )
        
        embed.add_field(
            name="🔧 Configuration",
            value=(
                "Set these in your `.env` file:\n"
                "• `EXTRALIFE_TEAM_ID`\n"
                "• `EXTRALIFE_PARTICIPANT_ID`\n"
                "Stats auto-update every 5 minutes"
            ),
            inline=False
        )
        
        await interaction.response.send_message(embed=embed, ephemeral=True)

async def setup(bot):
    """Setup function for the cog."""
    await bot.add_cog(ExtraLife(bot))

File: dist (# Name clash 2025-07-11 1utwh6C #)/commands/extralife.js
"use strict";
/**
 * Extra-Life charity integration commands
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extraLifeCommands = void 0;
exports.handleExtraLifeCommand = handleExtraLifeCommand;
exports.handleExtraLifeHelpCommand = handleExtraLifeHelpCommand;
exports.startExtraLifeUpdates = startExtraLifeUpdates;
const discord_js_1 = require("discord.js");
const axios_1 = __importDefault(require("axios"));
const node_cron_1 = __importDefault(require("node-cron"));
const config_1 = require("../config");
const database_1 = require("../utils/database");
const logger_1 = require("../utils/logger");
const dataManager = new database_1.DataManager();
let announcementChannels = new Map();
let pinnedMessages = new Map();
exports.extraLifeCommands = [
    new discord_js_1.SlashCommandBuilder()
        .setName('extralife')
        .setDescription('Extra-Life charity integration commands')
        .addStringOption(option => option.setName('action')
        .setDescription('Action to perform')
        .setRequired(true)
        .addChoices({ name: 'Show Stats', value: 'stats' }, { name: 'Setup Channel', value: 'setup' }, { name: 'Post Announcement', value: 'announce' }, { name: 'Refresh Data', value: 'refresh' }))
        .addChannelOption(option => option.setName('channel')
        .setDescription('Channel for announcements (admin only)')
        .setRequired(false)),
    new discord_js_1.SlashCommandBuilder()
        .setName('extralife-help')
        .setDescription('Show Extra-Life integration help')
];
async function handleExtraLifeCommand(interaction) {
    try {
        const options = interaction.options;
        const action = options.get('action')?.value;
        const channel = options.get('channel')?.channel;
        switch (action) {
            case 'stats':
                await showStats(interaction);
                break;
            case 'setup':
                await setupAnnouncements(interaction, channel);
                break;
            case 'announce':
                await postAnnouncement(interaction);
                break;
            case 'refresh':
                await refreshStats(interaction);
                break;
            default:
                await interaction.reply({
                    content: '❌ Invalid action. Use: `stats`, `setup`, `announce`, or `refresh`',
                    ephemeral: true
                });
        }
    }
    catch (error) {
        logger_1.logger.error('Error in extralife command:', error);
        const errorMessage = error instanceof Error ? error.message : 'Unknown error occurred';
        if (!interaction.replied) {
            await interaction.reply({
                content: `❌ Error: ${errorMessage}`,
                ephemeral: true
            });
        }
    }
}
async function handleExtraLifeHelpCommand(interaction) {
    const embed = new discord_js_1.EmbedBuilder()
        .setTitle('🎮 Extra-Life Integration Help')
        .setDescription('Commands for charity event support')
        .setColor('#0099ff')
        .addFields({
        name: '📊 `/extralife action:stats`',
        value: 'Show current fundraising statistics',
        inline: false
    }, {
        name: '⚙️ `/extralife action:setup`',
        value: 'Setup announcement channel (admin only)',
        inline: false
    }, {
        name: '📢 `/extralife action:announce`',
        value: 'Post event announcement with current stats',
        inline: false
    }, {
        name: '🔄 `/extralife action:refresh`',
        value: 'Manually refresh statistics from API',
        inline: false
    }, {
        name: '🔧 Configuration',
        value: [
            'Set these in your `.env` file:',
            '• `EXTRALIFE_TEAM_ID`',
            '• `EXTRALIFE_PARTICIPANT_ID`',
            'Stats auto-update every 5 minutes'
        ].join('\n'),
        inline: false
    });
    await interaction.reply({ embeds: [embed], ephemeral: true });
}
async function showStats(interaction) {
    await interaction.deferReply();
    try {
        // Try to get cached data first
        let data = await dataManager.loadExtraLifeCache();
        if (!data) {
            data = await fetchExtraLifeData();
            if (data) {
                await dataManager.saveExtraLifeCache(data);
            }
        }
        if (!data) {
            await interaction.followUp({
                content: '❌ Unable to fetch Extra-Life data. Check configuration.',
                ephemeral: true
            });
            return;
        }
        const embed = createStatsEmbed(data);
        await interaction.followUp({ embeds: [embed] });
    }
    catch (error) {
        logger_1.logger.error('Error showing stats:', error);
        await interaction.followUp({
            content: `❌ Error fetching stats: ${error}`,
            ephemeral: true
        });
    }
}
async function setupAnnouncements(interaction, channel) {
    // Check permissions
    if (!interaction.memberPermissions?.has(discord_js_1.PermissionFlagsBits.Administrator)) {
        await interaction.reply({
            content: '❌ Administrator permissions required.',
            ephemeral: true
        });
        return;
    }
    const targetChannel = channel || interaction.channel;
    const guildId = interaction.guildId;
    announcementChannels.set(guildId, targetChannel);
    // Save to database
    const settings = await dataManager.loadBotSettings(guildId);
    settings.extraLifeChannelId = targetChannel.id;
    await dataManager.saveBotSettings(guildId, settings);
    const embed = new discord_js_1.EmbedBuilder()
        .setTitle('🎮 Extra-Life Setup Complete')
        .setDescription(`Announcements will be posted in ${targetChannel}`)
        .setColor('#00ff00');
    await interaction.reply({ embeds: [embed] });
}
async function postAnnouncement(interaction) {
    const guildId = interaction.guildId;
    let targetChannel = announcementChannels.get(guildId);
    if (!targetChannel) {
        // Try to load from database
        const settings = await dataManager.loadBotSettings(guildId);
        if (settings.extraLifeChannelId) {
            const channel = interaction.guild?.channels.cache.get(settings.extraLifeChannelId);
            if (channel) {
                targetChannel = channel;
                announcementChannels.set(guildId, channel);
            }
        }
    }
    if (!targetChannel) {
        await interaction.reply({
            content: '❌ No announcement channel set. Use `/extralife action:setup` first.',
            ephemeral: true
        });
        return;
    }
    await interaction.deferReply();
    try {
        const data = await fetchExtraLifeData();
        if (!data) {
            await interaction.followUp({
                content: '❌ Unable to fetch Extra-Life data.',
                ephemeral: true
            });
            return;
        }
        const embed = createAnnouncementEmbed(data);
        // Post to announcement channel
        const message = await targetChannel.send({ embeds: [embed] });
        // Try to pin the message
        try {
            await message.pin();
            pinnedMessages.set(guildId, message);
        }
        catch (error) {
            logger_1.logger.warn('Could not pin announcement message:', error);
        }
        await interaction.followUp({
            content: `✅ Announcement posted in ${targetChannel}`
        });
    }
    catch (error) {
        logger_1.logger.error('Error posting announcement:', error);
        await interaction.followUp({
            content: `❌ Error posting announcement: ${error}`,
            ephemeral: true
        });
    }
}
async function refreshStats(interaction) {
    await interaction.deferReply();
    try {
        const data = await fetchExtraLifeData();
        if (data) {
            await dataManager.saveExtraLifeCache(data);
            const embed = createStatsEmbed(data);
            await interaction.followUp({
                content: '✅ Stats refreshed!',
                embeds: [embed]
            });
        }
        else {
            await interaction.followUp({
                content: '❌ Unable to fetch fresh data.',
                ephemeral: true
            });
        }
    }
    catch (error) {
        logger_1.logger.error('Error refreshing stats:', error);
        await interaction.followUp({
            content: `❌ Error refreshing: ${error}`,
            ephemeral: true
        });
    }
}
async function fetchExtraLifeData() {
    if (!config_1.config.hasExtraLifeConfig) {
        logger_1.logger.warn('No Extra-Life configuration found');
        return null;
    }
    const data = {};
    try {
        // Fetch team data
        if (config_1.config.extraLifeTeamUrl) {
            const teamResponse = await axios_1.default.get(config_1.config.extraLifeTeamUrl, {
                timeout: 10000,
                headers: { 'User-Agent': 'Dune-Discord-Bot/1.0' }
            });
            if (teamResponse.status === 200) {
                data.team = teamResponse.data;
            }
        }
        // Fetch participant data
        if (config_1.config.extraLifeParticipantUrl) {
            const participantResponse = await axios_1.default.get(config_1.config.extraLifeParticipantUrl, {
                timeout: 10000,
                headers: { 'User-Agent': 'Dune-Discord-Bot/1.0' }
            });
            if (participantResponse.status === 200) {
                data.participant = participantResponse.data;
            }
        }
        return Object.keys(data).length > 0 ? data : null;
    }
    catch (error) {
        logger_1.logger.error('Error fetching Extra-Life data:', error);
        return null;
    }
}
function createStatsEmbed(data) {
    const embed = new discord_js_1.EmbedBuilder()
        .setTitle('🎮 Extra-Life Statistics')
        .setColor('#0099ff')
        .setTimestamp();
    // Team stats
    if (data.team) {
        const team = data.team;
        embed.addFields({
            name: '👥 Team Stats',
            value: [
                `**Name:** ${team.name || 'Unknown'}`,
                `**Goal:** $${(team.fundraisingGoal || 0).toLocaleString()}`,
                `**Raised:** $${(team.sumDonations || 0).toLocaleString()}`,
                `**Members:** ${team.numMembers || 0}`
            ].join('\n'),
            inline: true
        });
    }
    // Participant stats
    if (data.participant) {
        const participant = data.participant;
        embed.addFields({
            name: '🎯 Participant Stats',
            value: [
                `**Name:** ${participant.displayName || 'Unknown'}`,
                `**Goal:** $${(participant.fundraisingGoal || 0).toLocaleString()}`,
                `**Raised:** $${(participant.sumDonations || 0).toLocaleString()}`,
                `**Donors:** ${participant.numDonations || 0}`
            ].join('\n'),
            inline: true
        });
    }
    // Progress calculation
    if (data.team) {
        const team = data.team;
        const goal = team.fundraisingGoal || 1;
        const raised = team.sumDonations || 0;
        const progress = Math.min(100, (raised / goal) * 100);
        embed.addFields({
            name: '📊 Progress',
            value: `**${progress.toFixed(1)}%** of goal reached`,
            inline: false
        });
    }
    embed.setFooter({ text: 'Data from Extra-Life API • Updates every 5 minutes' });
    return embed;
}
function createAnnouncementEmbed(data) {
    const embed = new discord_js_1.EmbedBuilder()
        .setTitle('🎮 Extra-Life Charity Event!')
        .setDescription([
        'Join us in supporting Children\'s Miracle Network Hospitals!',
        'Every donation helps kids in our local hospitals.'
    ].join(' '))
        .setColor('#ffd700');
    if (data.team) {
        const team = data.team;
        embed.addFields({ name: '🎯 Our Goal', value: `$${(team.fundraisingGoal || 0).toLocaleString()}`, inline: true }, { name: '💰 Raised So Far', value: `$${(team.sumDonations || 0).toLocaleString()}`, inline: true }, { name: '👥 Team Members', value: `${team.numMembers || 0} heroes`, inline: true });
    }
    embed.addFields({
        name: '🎮 How to Help',
        value: [
            '• Donate directly to our team',
            '• Share our fundraising page',
            '• Join our gaming events',
            '• Spread the word!'
        ].join('\n'),
        inline: false
    });
    const teamId = config_1.config.getConfig().extraLifeTeamId;
    if (teamId) {
        const teamPage = `https://www.extra-life.org/team/${teamId}`;
        embed.addFields({
            name: '🔗 Donation Link',
            value: `[Visit Our Team Page](${teamPage})`,
            inline: false
        });
    }
    embed.setFooter({ text: 'Thank you for supporting children\'s hospitals! 💙' });
    return embed;
}
// Background task to update statistics
function startExtraLifeUpdates() {
    if (!config_1.config.hasExtraLifeConfig) {
        logger_1.logger.info('Extra-Life configuration not found, skipping background updates');
        return;
    }
    // Update every 5 minutes
    node_cron_1.default.schedule('*/5 * * * *', async () => {
        try {
            logger_1.logger.debug('Running Extra-Life background update');
            const data = await fetchExtraLifeData();
            if (data) {
                await dataManager.saveExtraLifeCache(data);
                // Update pinned messages
                for (const [guildId, message] of pinnedMessages) {
                    try {
                        const embed = createAnnouncementEmbed(data);
                        await message.edit({ embeds: [embed] });
                    }
                    catch (error) {
                        logger_1.logger.warn(`Failed to update pinned message for guild ${guildId}:`, error);
                        pinnedMessages.delete(guildId);
                    }
                }
            }
        }
        catch (error) {
            logger_1.logger.error('Error in Extra-Life background update:', error);
        }
    });
    logger_1.logger.info('Extra-Life background updates started (every 5 minutes)');
}
